fw/rbcx-coprocessor/src/UltrasoundController.cpp

```diff
@@ -0,0 +1,94 @@
+#include "UltrasoundController.hpp"
+#include "stm32f1xx_hal.h"
+#include "stm32f1xx_hal_rcc.h"
+#include "stm32f1xx_ll_exti.h"
+#include "stm32f1xx_ll_tim.h"
+#include <optional>
+
+#include "Bsp.hpp"
+#include "Dispatcher.hpp"
+
+static const uint16_t pingTimeoutMicros = 30'000;
+static CoprocStat status;
+static std::optional<int> utsActiveIndex;
+static uint16_t risingEdgeMicros;
+
+void ultrasoundInit() {
+    LL_TIM_InitTypeDef timInit;
+    LL_TIM_StructInit(&timInit);
+    auto apb1TimClk = 2 * HAL_RCC_GetPCLK1Freq();
+
+    timInit.Autoreload = pingTimeoutMicros;
+    timInit.Prescaler = apb1TimClk / 1'000'000; // 1 us ticks
+    LL_TIM_Init(utsTimer, &timInit);
+    LL_TIM_SetOnePulseMode(utsTimer, LL_TIM_ONEPULSEMODE_SINGLE);
+    LL_TIM_SetUpdateSource(utsTimer, LL_TIM_UPDATESOURCE_COUNTER);
+    LL_TIM_ClearFlag_UPDATE(utsTimer);
+
+    LL_EXTI_DisableIT_0_31(utsEchoPins.second);
+    HAL_NVIC_SetPriority(utsTimerIRQn, utsIRQPrio, 0);
+    HAL_NVIC_EnableIRQ(utsTimerIRQn);
+}
+
+void ultrasoundDispatch(const CoprocReq_UltrasoundReq& request) {
+    if (request.utsIndex > 3) {
+        return;
+    }
+
+    switch (request.which_utsCmd) {
+    case CoprocReq_UltrasoundReq_singlePing_tag:
+        LL_TIM_DisableIT_UPDATE(utsTimer);
+        LL_EXTI_DisableIT_0_31(utsEchoPins.second);
+        utsActiveIndex = request.utsIndex;
+        LL_TIM_GenerateEvent_UPDATE(utsTimer);
+        LL_TIM_EnableCounter(utsTimer);
+
+        pinWrite(utsTrigPin[request.utsIndex], true);
+        while (LL_TIM_GetCounter(utsTimer) < 10
```

Comment:
Busy wait v hlavní smyčce? :C Nemám představu, kolik je 10us, je to "dostatečně málo"?

Reply:
Hele aktuálně je to 480 cyklů. Za mě je to odhadem OK. Ten systém přeci jen není jen to jádro - DMA přenosy, příjem, vysílání pořád běží :)

Reply:
To jsou tak dvě násobení nebo dělení s floatem, tak to by mohlo být ok)